src/Compilers/Server/VBCSCompilerTests/CompilerServerTests.cs

```diff
@@ -143,39 +98,117 @@ private static void CheckForBadShared(string arguments)
+            finally
+            {
+                foreach (var variable in resetVariables)
+                {
+                    Environment.SetEnvironmentVariable(variable.Key, variable.Value);
+                }
+            }
+        }
+
+        private static (T result, string output) UseTextWriter<T>(Encoding encoding, Func<TextWriter, T> func)
+        {
+            MemoryStream memoryStream;
+            TextWriter writer;
+            if (encoding == null)
+            {
+                memoryStream = null;
+                writer = new StringWriter();
+            }
+            else
+            {
+                memoryStream = new MemoryStream();
+                writer = new StreamWriter(memoryStream, encoding);
+            }
+            var result = func(writer);
+            writer.Flush();
+            if (memoryStream != null)
+            {
+                return (result, encoding.GetString(memoryStream.GetBuffer(), 0, (int)memoryStream.Length));
+            }
+            else
+            {
+                return (result, ((StringWriter)writer).ToString());
+            }
+        }
+
+        internal static (int exitCode, string output) RunCommandLineCompiler(
+            RequestLanguage language,
+            string argumentsSingle,
+            TempDirectory currentDirectory,
+            IEnumerable<KeyValuePair<string, string>> filesInDirectory = null,
+            IEnumerable<KeyValuePair<string, string>> additionalEnvironmentVars = null,
+            Encoding redirectEncoding = null,
+            bool shouldRunOnServer = true)
+        {
+            var arguments = new List<string>(argumentsSingle.Split(' '));
+            ReferenceNetstandardDllIfCoreClr(arguments);
+            CheckForBadShared(arguments);
+            CreateFiles(currentDirectory, filesInDirectory);
+
+            // Create a client to run the build.  Infinite timeout is used to account for the
+            // case where these tests are run under extreme load.  In high load scenarios the
+            // client will correctly drop down to a local compilation if the server doesn't respond
+            // fast enough.
+            var client = ServerUtil.CreateBuildClient(language);
+            client.TimeoutOverride = Timeout.Infinite;
+
+#if NET46
+            var sdkDir = RuntimeEnvironment.GetRuntimeDirectory();
+#else
+            string sdkDir = null;
+#endif
+
+            var buildPaths = new BuildPaths(
+                clientDir: Path.GetDirectoryName(typeof(CommonCompiler).Assembly.Location),
+                workingDir: currentDirectory.Path,
+                sdkDir: sdkDir,
+                tempDir: BuildServerConnection.GetTempPath(currentDirectory.Path));
+
+            var (result, output) = UseTextWriter(redirectEncoding, writer => ApplyEnvironmentVariables(additionalEnvironmentVars, () => client.RunCompilation(arguments, buildPaths, writer)));
+            Assert.Equal(shouldRunOnServer, result.RanOnServer);
```

Comment:
Will this `Assert` fail in high load scenarios where a local compilation was used?

Reply:
No, we wait forever to make sure we use the server - look at the change just above this (which was copied from existing code) https://github.com/dotnet/roslyn/pull/22291/files/2e29b1b669b8e3e0f166559e1d9b9fa5fb4e510f#diff-081b29b50ba2ad8251edc4c523a5af69R190

// Create a client to run the build.  Infinite timeout is used to account for the case where these tests are run under extreme load.  In high load scenarios the client will correctly drop down to a local compilation if the server doesn't respond fast enough.
```
client.TimeoutOverride = Timeout.Infinite;
```

Reply:
I see, thanks. I had misunderstood that comment.

Reply:
[Blame Jared's wording :)](https://github.com/khyperia/roslyn/commit/f8a981be0a7e3690f507044cd308b3151a079afd#diff-081b29b50ba2ad8251edc4c523a5af69R947)